osu.Game/Rulesets/Mods/IApplicableFailOverride.cs

```diff
@@ -18,5 +18,10 @@ public interface IApplicableFailOverride : IApplicableMod
         /// Whether we want to restart on fail. Only used if <see cref="PerformFail"/> returns true.
         /// </summary>
         bool RestartOnFail { get; }
+
+        /// <summary>
+        /// Whether mod triggered the fail.
+        /// </summary>
+        bool TriggeredFail { get; }
```

Comment:
This design is back-to-front. That sort of gameplay-state-storing field cannot exist on a mod. The machinery in `HealthProcessor` should be responsible for determining which particular mod(s) triggered the failure and using that knowledge internally to determine which mod effects are eligible to be applied.